fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -238,4 +256,79 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
+        if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
+            return fileDescriptor;
+        }
+        DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
+        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
+        fileBuilder.addMessageType(MetaDataProtoEditor.createDefaultUnion(fileDescriptor));
+        try {
+            return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
+        } catch (Descriptors.DescriptorValidationException e) {
+            throw new MetaDataException("Failed to add a default union", e);
+        }
+    }
+
+    @Nonnull
+    private static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.Descriptor messageType : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                addFieldToUnion(unionMessageType, messageType.getName(), messageType.getFullName());
+            }
+        }
+        return unionMessageType;
+    }
+
+    /**
+     * Creates a default union descriptor for the given file descriptor and a base union descriptor. It adds all of the
+     * non-{@code NESTED} message types that exist in the base union to the synthetic union.
+     * @param fileDescriptor the file descriptor to create a union for
+     * @param baseUnionDescriptor the base union descriptor
+     * @return the builder for the union
+     */
+    @Nonnull
+    @API(API.Status.INTERNAL)
+    public static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull Descriptors.Descriptor baseUnionDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.FieldDescriptor field : baseUnionDescriptor.getFields()) {
+            Descriptors.Descriptor messageType = fileDescriptor.findMessageTypeByName(field.getMessageType().getName());
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                addFieldToUnion(unionMessageType, messageType.getName(), messageType.getFullName(), field.getNumber(), field.getOptions());
```

Comment:
Is there a reason this can't be `unionMessageType.addField(field.toProto().toBuilder().setTypeName(messageType.getFullName())`? Then I think one could clean up `addFieldToUnion` to not take as many parameters. (I think one could also remove it from this file and then it doesn't need to be `public`, though arguably its a clearer separation of responsibilities here.)

Reply:
I just wanted to reuse code, I can split them too. 

I was thinking that this class can be the eventual host of all meta-data editing duties, so, it would be good keep it here.